### ISSUE TITLE
Training nerf with less levels

### ISSUE THREAD

**Issue by MiZhenxing** [NVlabs/instant-ngp#140]
Hi, thank you for your great project.

I would like to explore the properties of your project with extreme level numbers. So I have trained nerf with several configs, with configs and screenshots below:

I first tried with "n_levels": 1, "n_features_per_level": 8
config 1:
```
{
	"parent" : "base.json",
	"encoding": {
		"otype": "HashGrid",
		"n_levels": 1,
		"n_features_per_level": 8,
		"log2_hashmap_size": 19,
		"base_resolution": 16
	}
}
```
screenshot 1:

![image](https://user-images.githubusercontent.com/34396974/153806245-73e27ee4-e69d-44b5-a7dd-550574f3c485.png)
We got nan loss after several steps.
Then I tried to change the "base_resolution" to [16, 2048] with "n_levels": 1, I always got nan loss.

So I changed to "n_levels": 2
config 2:

```
{
	"parent" : "base.json",
	"encoding": {
		"otype": "HashGrid",
		"n_levels": 2,
		"n_features_per_level": 8,
		"log2_hashmap_size": 19,
		"base_resolution": 16
	}
}
```
screenshot 2:
![image](https://user-images.githubusercontent.com/34396974/153806411-3e5d528a-9752-4b73-983f-923d69c01f86.png)
The training continues but the result seems coarse.

Then I changed to "n_levels": 4
config 3:
```
{
	"parent" : "base.json",
	"encoding": {
		"otype": "HashGrid",
		"n_levels": 4,
		"n_features_per_level": 8,
		"log2_hashmap_size": 19,
		"base_resolution": 16
	}
}
```
screenshot 3:
![image](https://user-images.githubusercontent.com/34396974/153806836-a296ebf4-8617-4101-b514-2686a2ec24b4.png)
and get pretty reasonable results:

After that, I tried to  change the "base_resolution" to [16, 2048] with "n_levels": 2,
config 4:
```
{
	"parent" : "base.json",
	"encoding": {
		"otype": "HashGrid",
		"n_levels": 2,
		"n_features_per_level": 8,
		"log2_hashmap_size": 19,
		"base_resolution": 64
	}
}
```
screenshot 4:

![image](https://user-images.githubusercontent.com/34396974/153807688-6552703a-b728-4bec-a42c-9d59f31921ce.png)

The result became finer.

Then with "base_resolution": 128
config 5:
```
{
	"parent" : "base.json",
	"encoding": {
		"otype": "HashGrid",
		"n_levels": 2,
		"n_features_per_level": 8,
		"log2_hashmap_size": 19,
		"base_resolution": 128
	}
}
```
screenshot 5:
![image](https://user-images.githubusercontent.com/34396974/153807810-8880b606-cf82-4dae-84f7-cf298a532898.png)

There was more high-frequency noise.  After longer training, the result became reasonable.
screenshot 6:
![image](https://user-images.githubusercontent.com/34396974/153809991-4cab0250-a9cb-4309-ac27-498c7dd8cfc2.png)
But with more challenging fox dataset, the results with config 5 seems not as reasonable even with longer training.
screenshot 7:
![image](https://user-images.githubusercontent.com/34396974/153810547-a711b979-aac2-4cec-a004-eb81a3b95bf6.png)

So I would like to ask how could I successfully train a network with "n_levels": 1. Have you trained such networks? Also could you elaborate more about how the hyperparameters such n_levels and base_resolution infulence the final results?

Thank you very much!

**Comment by resurgo97** [comment]
When "n_levels" is 1, it becomes not multi-resolutional anymore.
My assumption is you should increase T to compensate the loss of the representative power.
I guess then you will lose the efficiency that comes from having multi-resolution though.

**Comment by MiZhenxing** [comment]
> When "n_levels" is 1, it becomes not multi-resolutional anymore. My assumption is you should increase T to compensate the loss of the representative power. I guess then you will lose the efficiency that comes from having multi-resolution though.

Hi, thank you for your suggestion. I also increased T from 19 to 24 with level 1 and also got nan loss. I think maybe the problem is not related to the representative power of the model. It may be related to the implementation. The implementation may not explicitly handle the situation of level 1.

**Comment by JoanCharmant** [comment]
If you have only one level with base resolution 16, it means you are training a 16³ grid (4096 voxels), so it already largely fits into the 2¹⁹ hash table (it won't even use the hashing algorithm in that case) and increasing T won't change anything in that case. For a single level you should use a much finer grid. 

Then since there is only one level any hash collision will be very bad, creating floaters from voxels in empty space sharing their features with voxels on the surface, so T should be increased as well.

**Comment by Tom94** [comment]
Hi there, it looks like I introduced a padding-related bug in one of the recent optimizations.

I've just pushed a fix here: https://github.com/NVlabs/tiny-cuda-nn/commit/7553d301ffafec4c341dba5355606dcf4102b9be

Please pull the latest `master` and make sure to update submodules `git submodule update --recursive`. This should fix the NaN loss & improve quality of `n_levels`<16

**Comment by MiZhenxing** [comment]
> Hi there, it looks like I introduced a padding-related bug in one of the recent optimizations.
> 
> I've just pushed a fix here: [NVlabs/tiny-cuda-nn@7553d30](https://github.com/NVlabs/tiny-cuda-nn/commit/7553d301ffafec4c341dba5355606dcf4102b9be)
> 
> Please pull the latest `master` and make sure to update submodules `git submodule update --recursive`. This should fix the NaN loss & improve quality of `n_levels < 16`

Thank you! I have tested the new version and I no longer get nan loss.
